aggregator/flush_mgr.go

```diff
@@ -0,0 +1,338 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/sync"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errFlushManagerClosed = errors.New("flush manager is closed")
+)
+
+// PeriodicFlusher flushes periodically.
+type PeriodicFlusher interface {
+	// FlushInterval returns the periodic flush interval.
+	FlushInterval() time.Duration
+
+	// Flush performs a flush.
+	Flush()
+}
+
+// FlushManager manages and coordinates flushing activities across many
+// periodic flushers with different flush intervals with controlled concurrency
+// for flushes to minimize spikes in CPU load and reduce p99 flush latencies.
+type FlushManager interface {
+	// Register registers a metric list with the flush manager.
+	Register(flusher PeriodicFlusher) error
+
+	// Close closes the flush manager.
+	Close()
+}
+
+type flushManagerMetrics struct {
+	queueSize tally.Gauge
+}
+
+func newFlushManagerMetrics(scope tally.Scope) flushManagerMetrics {
+	return flushManagerMetrics{
+		queueSize: scope.Gauge("queue-size"),
+	}
+}
+
+type flushManager struct {
+	sync.Mutex
+
+	nowFn         clock.NowFn
+	checkEvery    time.Duration
+	jitterEnabled bool
+	workers       xsync.WorkerPool
+	scope         tally.Scope
+
+	closed     bool
+	rand       *rand.Rand
+	buckets    []*flushBucket
+	flushTimes flushMetadataHeap
+	sleepFn    sleepFn
+	wgFlush    sync.WaitGroup
+	metrics    flushManagerMetrics
+}
+
+// NewFlushManager creates a new flush manager.
+func NewFlushManager(opts FlushManagerOptions) FlushManager {
+	if opts == nil {
+		opts = NewFlushManagerOptions()
+	}
+	nowFn := opts.ClockOptions().NowFn()
+	scope := opts.InstrumentOptions().MetricsScope()
+	rand := rand.New(rand.NewSource(nowFn().UnixNano()))
+	workers := xsync.NewWorkerPool(opts.WorkerPoolSize())
+	workers.Init()
+
+	mgr := &flushManager{
+		nowFn:         nowFn,
+		checkEvery:    opts.CheckEvery(),
+		jitterEnabled: opts.JitterEnabled(),
+		workers:       workers,
+		scope:         scope,
+		rand:          rand,
+		sleepFn:       time.Sleep,
+		metrics:       newFlushManagerMetrics(scope),
+	}
+	if mgr.checkEvery > 0 {
+		mgr.wgFlush.Add(1)
+		go mgr.flush()
+	}
+	return mgr
+}
+
+func (mgr *flushManager) Register(flusher PeriodicFlusher) error {
+	mgr.Lock()
+	bucket, err := mgr.bucketForWithLock(flusher)
+	if err == nil {
+		bucket.Add(flusher)
+		mgr.Unlock()
+		return nil
+	}
+	mgr.Unlock()
+	return err
+}
+
+func (mgr *flushManager) bucketForWithLock(l PeriodicFlusher) (*flushBucket, error) {
+	if mgr.closed {
+		return nil, errFlushManagerClosed
+	}
+	flushInterval := l.FlushInterval()
+	for _, bucket := range mgr.buckets {
+		if bucket.interval == flushInterval {
+			return bucket, nil
+		}
+	}
+	bucketScope := mgr.scope.SubScope("bucket").Tagged(map[string]string{
+		"interval": flushInterval.String(),
```

Comment:
`bucket` is already in the name of the metric, I can change the tag name to `flush-interval` if that's better?